geoh5py/shared/entity_type.py

```diff
@@ -20,136 +20,210 @@
-        self._workspace = weakref.ref(workspace)
-
-        assert uid is None or isinstance(uid, uuid.UUID)
-        self._description: str | None = "Entity"
-        self._name: str | None = "Entity"
-        self._on_file = False
-        self._uid: uuid.UUID = uid if uid is not None else uuid.uuid4()
+    # pylint: disable=too-many-arguments
+    """
+    The base class for all entity types.
+
+    :param workspace: The workspace to associate the entity type with.
+    :param uid: The unique identifier of the entity type.
+    :param description: The description of the entity type.
+    :param name: The name of the entity type.
+    :param entity_class: The class of the entity.
+    :param on_file: Return True if the entity is on file.
+    """
 
-        for attr, item in kwargs.items():
-            try:
-                if attr in self._attribute_map:
-                    attr = self._attribute_map[attr]
-                setattr(self, attr, item)
-            except AttributeError:
-                continue
+    _attribute_map = {"Description": "description", "ID": "uid", "Name": "name"}
 
-        self.workspace.register(self)
+    def __init__(
+        self,
+        workspace: Workspace,
+        uid: uuid.UUID | None = None,
+        description: str | None = "Entity",
+        name: str | None = "Entity",
+        entity_class: type | None = None,
+        on_file: bool = False,
+    ):
+        self._uid: uuid.UUID = ensure_uuid(uid) if uid is not None else uuid.uuid4()
+
+        self.description = description
+        self.name = name
+        self.entity_class = entity_class
+        self.workspace = workspace
+        self.on_file = on_file
 
     @property
-    def attribute_map(self):
+    def attribute_map(self) -> dict[str, str]:
         """
-        :obj:`dict` Correspondence map between property names used in geoh5py and
+        Correspondence map between property names used in geoh5py and
         geoh5.
         """
         return self._attribute_map
 
+    @classmethod
+    def convert_kwargs(cls, kwargs: dict[str, Any]) -> dict[str, Any]:
+        """
+        Convert the kwargs to the geoh5py attribute names.
+
+        :param kwargs: The kwargs to convert.
+
+        :return: The converted kwargs.
+        """
+        return {
+            cls._attribute_map.get(key, key): value for key, value in kwargs.items()
+        }
+
+    def copy(self, **kwargs) -> EntityType:
+        """
+        Copy this entity type to another workspace.
+        """
+
+        attributes = {
+            prop: getattr(self, prop)
+            for prop in dir(self)
+            if isinstance(getattr(self.__class__, prop, None), property)
+            and getattr(self, prop) is not None
+            and prop != "attribute_map"
+        }
+
+        attributes.update(kwargs)
+
+        if attributes.get("uid") in getattr(
+            attributes.get("workspace", self.workspace), "_types"
+        ):
+            del attributes["uid"]
```

Comment:
When would we copy the entity_type by itself? The copy should happen with the object and/or data entity no?

Reply:
I did not write this logic in the first place. 
It seems to me it's a shared function:

data_type can be copy in the case of a referenced data type. 
group_type and object_type can be copied too aren't they?

At last, during the copying, the object is copied except the changed attributes.
And the uid, of the object is not copied itself.

What is wrong with this logic?

Reply:
Yea, I am not sure. I guess let's leave it until something comes up.

Reply:
not touching it